### DISCUSSION TITLE
Injectable configuration

### DISCUSSION THREAD

**Discussion by Emily-Jiang** [eclipse-ee4j/config#111] [General]
Do we want to support Injectable configuration in this spec?

**Comment by radcortez** [comment]
No. It should be in the scope of CDI, not here.

**Reply by hantsy** [reply]
No, CDI is the kernel of Jakarta EE specs now, other specs should adapt CDI automatically.

Like the Spring IOC, all spring components are based on it.

**Comment by dmlloyd** [comment]
A qualified "yes" - injection as @radcortez points out is the purview of CDI, not us. However, it should be possible for an application to inject their configuration object (see my comments on [the "Programmatic API" thread](#110)) by its type, which should uniquely identify the application configuration. This specification should state that compliant implementations of this specification which _also_ implement the CDI specification are required to facilitate this in an implementation-specific manner.

**Reply by dmlloyd** [reply]
Another highly salient point is that I think we would like CDI itself to be configurable using the common configuration mechanism - perhaps eventually as a replacement for `beans.xml` for example. This implies that the configuration (or, at least the CDI portion of the configuration) must be available at the time that CDI is bootstrapping, potentially also providing application configuration to things like portable extensions at this time.

This limits the level of interaction that CDI can have with configuration. For example, CDI would not be usable as a way to modify or customize application configuration if that same configuration must be usable for configuring a portable extension.

**Reply by radcortez** [reply]
While I agree that it should be possible to inject a configuration object, I'm not convinced that it should be defined in the Config specification. I believe that most of us agreed that it should be possible to configure CDI with Config. This creates a unique relationship compared to the other platform specifications, where both specs depend on each other.

It is unclear to me how we will express such a relationship. I'm hesitant to make it a big thing in the specification itself because it may give the wrong impression that you need CDI to use the Config mechanism (do we?) and scare away other implementors outside Jakarta land. Also, if we need to provide some API that needs CDI, we get into circular dependency land, which we can essily fix. Still, I'm starting to think it would be cleaner to specify Config injection in the CDI directly.

**Reply by ljnelson** [reply]
The way I see it:
* CDI is (definitionally) a [Jakarta Dependency Injection](https://jakarta.ee/specifications/dependency-injection/2.0/apidocs/jakarta/inject/package-summary.html#package.description) implementation.  (Spring is another; HK2 is still another; I think Guice is another; I'm sure there are many others.)
* Jakarta Dependency Injection is a "consumer-only" specification, i.e. it does not really speak about how "injectors" work, other than the bare minimum required, only about how injection points work, regardless of how an implementation satisfies those injection points
* If we say that Jakarta Config in most (**but, very importantly, _not all_**) situations depends on Jakarta Dependency Injection, doesn't this dodge the circularity?  That gives you injection points, `jakarta.inject.Inject`, the non-proxied `Singleton` scope, the `Provider` construct, and qualifiers.  There's probably not a whole lot more that we would need.  Maybe that's a partial way out?

(Note: I deliberately say "most…situations" and not "all…situations" because there can be very important restrictions on injection imposed by Jakarta Dependency Injection implementations (like CDI). For example, within a CDI portable extension, you cannot `@Inject` anything, even though CDI is a Jakarta Dependency Injection implementation, because the container is still coming up.  I'm guessing there are similar restrictions in other Jakarta Dependency Injection implementations: I would imagine, though I'm no expert, that, say, in Guice you probably can't inject things inside injectors either.  That's another discussion related, perhaps, but not identical to, #110, and should not be addressed directly here.)

**Reply by dmlloyd** [reply]
I don't see any reason to specify this in any greater complexity than:

> Conforming implementations of Jakarta Config which are provide a conforming implementation of Jakarta CDI *must* allow configuration root objects to be injected into applications as beans. (plus any necessary additional information about qualifiers etc.)

In particular we do not need to establish a dependency either way, nor do we need to specify any additional requirements on CDI itself.

**Reply by ljnelson** [reply]
If all that's required is injection, don't you think we could replace "CDI" with "Jakarta Dependency Injection" instead?  Might be even simpler.

**Reply by dmlloyd** [reply]
If that is the proper name of the specification *or* a normative term then yes.

**Reply by ljnelson** [reply]
For clarity (and the benefit of others): [Jakarta Dependency Injection](https://jakarta.ee/specifications/dependency-injection/2.0/) is its own spec.  [CDI](https://jakarta.ee/specifications/cdi/4.0/) is [an extension of it](https://jakarta.ee/specifications/cdi/4.0/jakarta-cdi-spec-4.0.html#_relationship_to_jakarta_dependency_injection).

(Weirdly the [CDI page](https://jakarta.ee/specifications/cdi/4.0/) calls itself Jakarta Contexts and Dependency Injection, Jakarta Context Dependency Injection (!) and Jakarta Contexts Dependency Injection (!).  Yikes.  I think the first one is the proper name for it.)

**Comment by ljnelson** [comment]
I see configuration as a spectrum:

![Configuration Spectrum](https://user-images.githubusercontent.com/80764/189167923-d7cfd697-89ff-4f32-9d65-ca7bb63be3d3.png)

(In terms of the spectrum, injecting configuration falls in the yellow/left end of the spectrum.)

**I agree that injection of configuration (where that term is deliberately left undefined here) should be supported where possible.**

I agree that CDI should be the mechanism by which injection is implemented wherever possible.

As David notes correctly, there are also cases where injection cannot be supported (from within a portable extension, for example).

**Comment by hantsy** [comment]
Absolutely yes.

**Reply by Emily-Jiang** [reply]
The injectable relies Jakarta Dependency Injection not CDI. Let's not mix two things together. Jakarta Dependency Injection is not part of CDI but depend on CDI.
